Yuhsuan/issue 597 contrast bias

YuHsuan-Hwang opened the pull request:
This PR is for feature https://github.com/CARTAvis/carta-frontend/issues/597. Also fixed issue https://github.com/CARTAvis/carta-frontend/issues/1356.
The color scaling line on the histogram can be turned on/off in the setting panel so that users can export images with or without the line.

YuHsuan-Hwang committed:
add bias contrast inputs; rearrange config order

YuHsuan-Hwang committed:
add 2Dboard

YuHsuan-Hwang committed:
add drag contraint

YuHsuan-Hwang committed:
add handle dragMove; add Collapse

YuHsuan-Hwang committed:
fix button width

YuHsuan-Hwang committed:
fix button width

YuHsuan-Hwang committed:
add colormap scaling in the histogram

YuHsuan-Hwang committed:
add reset button

YuHsuan-Hwang committed:
zoom histogram to min/max

YuHsuan-Hwang committed:
fix conflicts and merge dev

YuHsuan-Hwang committed:
apply to colorbar

YuHsuan-Hwang committed:
add Shared/BiasContrastSelectComponent

YuHsuan-Hwang committed:
minor changes

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/issue_597_contrast_bias

YuHsuan-Hwang committed:
minor changes

YuHsuan-Hwang committed:
add show colorscale setting

YuHsuan-Hwang committed:
rename variables

YuHsuan-Hwang committed:
avoid exporting color scaling data

YuHsuan-Hwang committed:
add semicolon

YuHsuan-Hwang commented:
Implementing a new mode for smoothed bias and contrast by using Glynn Rogers function. Will reopen the PR after this part is completed.

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/issue_597_contrast_bias

YuHsuan-Hwang committed:
add smoothed bias/contrast

YuHsuan-Hwang committed:
add click event on 2d board

YuHsuan-Hwang committed:
add smoothed bias contrast to shader

YuHsuan-Hwang committed:
adjust histogram click event

YuHsuan-Hwang committed:
minor change

YuHsuan-Hwang commented:
@veggiesaurus @kswang1029 @duidae 
A new mode for smoothed bias and contrast is added and set as default.

Some details:
* Switch the mode in the render configuration settings panels -> Colormap Scaling.
* Adjust bias and contrast by dragging the point or clicking in the rectangle.

Please check if everything works well in both two modes. Thanks.

veggiesaurus commented on src/utilities/math.ts:
```diff
@@ -76,37 +76,81 @@ export function getPercentiles(histogram: CARTA.IHistogram, ranks: number[]): nu
+}
+
+export function scaleValue(x: number, scaling: FrameScaling, alpha: number = 1000, gamma: number = 1.5, bias: number = 0, contrast: number = 1,
+    biasContrastMode: boolean = false, smoothedBias: number = 0, smoothedContrast: number = 0) {
```
can we not just use the same arguments (`bias` and `contrast` for both modes, but use `biasContrastMode` to determine which function to use?

YuHsuan-Hwang replied:
Combined the parameters.

veggiesaurus commented on src/utilities/math.ts:
```diff
@@ -76,37 +76,81 @@ export function getPercentiles(histogram: CARTA.IHistogram, ranks: number[]): nu
-            return x;
+            scaleValue = x;
+    }
+    if (biasContrastMode) {
```
since `biasContrastMode` is a boolean, I think it just be something like `useSmoothBiasContrast`

YuHsuan-Hwang replied:
Sure. Will change the naming.

veggiesaurus commented on src/utilities/math.ts:
```diff
@@ -76,37 +76,81 @@ export function getPercentiles(histogram: CARTA.IHistogram, ranks: number[]): nu
+        scaleValue = (errorFunction(scaleValue, c, smoothedBias) - offset) / denominator;
+    } else {
+        scaleValue = clamp(scaleValue - bias, 0, 1);
+        scaleValue = clamp((scaleValue - 0.5) * contrast + 0.5, 0, 1);
```
if we use `(contrast + 1)`, wouldn't we be able to use the same range and defaults for both bias/contrast modes?

veggiesaurus replied:
(or perhaps we could change the other definition to match this mode? :thinking: )

YuHsuan-Hwang replied:
Combined the parameters.

veggiesaurus commented on src/stores/RenderConfigStore.ts:
```diff
@@ -277,6 +304,42 @@ export class RenderConfigStore {
+        this.contrast = 1;
+    };
+
+    @action setSmoothedBias = (smoothedBias: number) => {
```
again, I think we should just use the same bias/contrast variables here. When we switch modes, we can automatically reset them to the correct defaults

YuHsuan-Hwang replied:
Combined the parameters.

veggiesaurus commented on src/stores/RenderConfigStore.ts:
```diff
@@ -56,11 +56,22 @@ export class RenderConfigStore {
     static readonly GAMMA_MAX = 2;
     static readonly ALPHA_MIN = 0.1;
     static readonly ALPHA_MAX = 1000000;
+    static readonly BIAS_MIN = -1;
```
it would be nice if we could just adjust our bias/contrast equations to a form such that all of these variables just have the same range of `[0, 1]`.

kswang1029 replied:
For the 2D box, wonder if we can apply a transformation internally so that the default position of the circle is at the center of the box?

YuHsuan-Hwang replied:
@veggiesaurus Sure, we can modify the functions for normal bias/contrast and combine the parameters. We may need to consider that the default value for normal bias and contrast will become 0.5 and 0.5, and this is not as straightforward as 0 and 1.

YuHsuan-Hwang replied:
@kswang1029 not sure how we can transform the position to the center

kswang1029 commented:
@YuHsuan-Hwang would be good to add "double-click" (in the 2D box) as reset bias and contrast.

YuHsuan-Hwang commented:
> @YuHsuan-Hwang would be good to add "double-click" (in the 2D box) as reset bias and contrast.

Sure, I will try to add the event.

YuHsuan-Hwang commented:
@veggiesaurus @kswang1029 A solution for combining the parameters and placing the default circle in the center of the box:
 - Set the ranges and default values of smoothed bias/contrast the same as normal bias/contrast.
 - Map smoothed bias from [-1, 1] to [0, 1] and apply Glynn Rogers function.
 - Separate smoothed contrast into two range: [0, 1] and [1, 2]. For range [1, 2], map the value from [1, 2] to [0, 1] and apply Glynn Rogers function. For range [0, 1], apply function `clamp((value - (1.0 - smoothedBias)) * contrast + (1.0 - smoothedBias), 0, 1)`.

Cons: When contrast is [0, 1], the function is not so straightforward.

YuHsuan-Hwang committed:
adjust scaling function to combine parameters; add double click event

YuHsuan-Hwang committed:
minor changes

YuHsuan-Hwang commented:
> @veggiesaurus @kswang1029 A solution for combining the parameters and placing the default circle in the center of the box:
> 
> * Set the ranges and default values of smoothed bias/contrast the same as normal bias/contrast.
> * Map smoothed bias from [-1, 1] to [0, 1] and apply Glynn Rogers function.
> * Separate smoothed contrast into two range: [0, 1] and [1, 2]. For range [1, 2], map the value from [1, 2] to [0, 1] and apply Glynn Rogers function. For range [0, 1], apply function `clamp((value - (1.0 - smoothedBias)) * contrast + (1.0 - smoothedBias), 0, 1)`.
> 
> Cons: When contrast is [0, 1], the function is not so straightforward.

Applied the new transformations to bias/contrast and added double click event in the 2D box.

veggiesaurus commented:
@kswang1029 don't you think the "smooth bias/contrast" setting should be in preferences, rather than the widget config? It's related to the render configuration, not the widget settings.

One minor issue: when switching the "show color scaling" setting on and off, it changes the graph zoom. This is rather confusing, because it only occurs when the graph's bounds are automated. I suspect this may confuse users

veggiesaurus commented:
@YuHsuan-Hwang the contrast/bias widget works nicely if you click and drag on the _middle_ of the circle, but it's jumpy if I click and drag on the edge of the point: 


https://user-images.githubusercontent.com/592504/115233271-f7e17600-a117-11eb-8346-df338ecb1d88.mp4


I think you just need to take the initial offset into account when moving the point.

YuHsuan-Hwang commented:
> @YuHsuan-Hwang the contrast/bias widget works nicely if you click and drag on the _middle_ of the circle, but it's jumpy if I click and drag on the edge of the point:
> 
>  Peek.2021-04-19.14-01.mp4 
> I think you just need to take the initial offset into account when moving the point.

I see. Will change this part.

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/issue_597_contrast_bias

YuHsuan-Hwang committed:
change histogram zooming; change settings

YuHsuan-Hwang committed:
move useSmoothedBiasContrast to preferenceStore

YuHsuan-Hwang committed:
adjust drag event

YuHsuan-Hwang committed:
minor changes

YuHsuan-Hwang commented:
@kswang1029 @veggiesaurus @duidae Please check the new changes:
* Moved the smoothed bias/contrast setting to the preferences dialog. Removed the show scaling function setting.
* Show the full range of the histogram instead of zooming to Min/Max.
* Set the opacity of the scaling function line to 0.5.
* Fixed the issue of dragging on the edge of the point.

kswang1029 reviewed (approved):
Feature-wise this looks good. 👍

veggiesaurus commented:
@YuHsuan-Hwang we will need to adjust the preferences schema in order to accommodate your new setting, won't we?

veggiesaurus committed:
Merge branch 'dev' into yuhsuan/issue_597_contrast_bias

veggiesaurus committed:
update preference schema to include smoothed bias/contrast toggle
